### ISSUE TITLE
testscript: how to improve wall clock performance of long scripts?

### ISSUE THREAD

**Issue by twpayne** [rogpeppe/go-internal#103]
This is a Cunningham's Law issue ("the best way to get the right answer on the internet is not to ask a question; it's to post the wrong answer."). I'm pretty sure that this is the wrong answer.

tl;dr my increasingly-long test scripts are taking an increasingly-long time to run (now up to 20s for some) and I want to speed up my tests.

`testscript` already runs scripts in parallel. This is fantastic, and much appreciated (as is testscript).

My testscripts are starting to look like this:

```
# checkpoint start

long, complicated test-specific setup bit

# checkpoint setup complete

bunch of basic tests that could be done in parallel

# checkpoint basic tests complete

bunch of more complex tests that only make sense if the all basic tests passed,
and could also be done in parallel

# checkpoint done
```

I suspect that the following dialog will occur:

Me: "I want my tests to run faster"
testscript: "Split your tests into separate tests so I can run them in parallel"
Me: "But what about the complicated setup bit and basic test bits? suddenly my tests have loads of hard-to-maintain duplication 😢"
testscript: "You do know that you can write your own commands for testscript by adding them to `Params.Cmds`, right?"
Me: "Yes, and I'm already doing this, but my setup bits are test-specific."
testscript: "This is starting to get out of scope."
Me: "Yes, and imagine a world where testscripts have their own flow language built into them where you can say `fork` where a whole bunch of following tests written in the same file execute in parallel and `join` where execution only continues if all the tests so far have succeeded".
testscript: "Yes, that would be lovely. Dataflow languages are also complex and way beyond the scope of what testscript can realistically provide."
Me: "Fine. I'll just generate my tests with `text/template` in `TestMain` and testscript will never know."

**Comment by myitcv** [comment]
Two alternatives (assuming that splitting up the testscript tests is the "right" answer, i.e. constraints allow):

* use [CUE](https://cuelang.org) (but then again I'm biased 😄) to template the tests
* use an approach similar to [`txtarutil`](https://github.com/govim/govim/tree/8c4bca5906595372d6afd4b97abe5b5fc272de9c/internal/cmd/txtarutil):

https://github.com/govim/govim/blob/8c4bca5906595372d6afd4b97abe5b5fc272de9c/cmd/govim/main_test.go#L30

**Comment by mvdan** [comment]
I have had similar questions in the past, and I ended up just making `testscript.Setup` do as much work as possible, even if it's not needed for some of the tests. It's nicer because I can write that in Go, and it's still fully parallel.

**Comment by twpayne** [comment]
Thanks for the suggestions and leads.

> * use [CUE](https://cuelang.org) (but then again I'm biased 😄) to template the tests

Ah, cool! I attended one of Marcel's talks on CUE a couple of years ago and it was already looking very promising then.

So, my current workaround is to use, er, none of the above. I took inspiration from @mvdan's approach (me taking inspiration from @mvdan is happening with alarming frequency at the moment), but as my tests sometimes require setup to be present, and sometimes not, I ended creating a bunch of setup commands that I can include or not in each `testscript`. This is as concise as using templates, but I still have the hassle of managing multiple files. It's OK for now though.

@myitcv, crazy idea because I'm not very familiar with CUE, but would it make sense to write the `testscript` scripts themselves in CUE? This would obviously mean creating quite a different package to `testscript`, but CUE's power brings many advantages. Note that previous attempts to use a language for representing data as a scripting language have been a disaster (I'm looking straight at you Ansible and your attempt to use YAML as a scripting language). Maybe CUE is better suited?

Anyhow, closing this issue. Many thanks for your fast responses :)

**Comment by myitcv** [comment]
> would it make sense to write the testscript scripts themselves in CUE?

Depends what you had in mind here. A testscript script is, after all, just a txtar archive. So there isn't much structure that would warrant a CUE schema.

The context for my suggesting CUE was actually to make templating quicker/easier than using `text/template`. For example, if in a number of tests you have a common block of commands, or want to share a golden file. Or you want to specify a multi-factor matrix of tests and have CUE generate the cross product of that matrix (I have a particular example in mind for [`govim`](https://github.com/govim/govim)).

Definitely not mutually exclusive with setup commands, and certainly not a substitute for doing some setup in Go (for example in `govim` we have a hierarchy of testscript directories with different environment setups for each, where that environment is established _before_ the script runs).

At the end of the day, just kicking round ideas 😄
